### ISSUE TITLE
[Feature]: Add more granular permissions and user roles.

### ISSUE THREAD

**Issue by DeVoresyah** [coollabsio/coolify#2378]
### Description

Currently, team members apply for the entire general system. I want to have a specific team member for each resource. So, imagine that each resource has a new menu called members / team / team members where the admin can invite people to specific resources.
![image](https://github.com/coollabsio/coolify/assets/8052370/7cf25b76-bfef-4c4c-96f6-3304f7360741)

### Minimal Reproduction (if possible, example repository)

n/a

### Version

n/a

**Comment by abhishek818** [comment]
@DeVoresyah can i pick this up ?
/attempt #2378

**Comment by DeVoresyah** [comment]
hi @abhishek818 , any update on this?

**Comment by WcaleNieWolny** [comment]
Hello, i would like to attempt this

**Comment by DeVoresyah** [comment]
> Hello, i would like to attempt this

yes you can

**Comment by WcaleNieWolny** [comment]
I see the following split of permissions possible:

![image](https://github.com/user-attachments/assets/4c072f89-1600-42d7-a495-d521cf875ade)

Would you like to implement only the `resource` layer? So that you can either
A) Add someone to the whole team
B) Add someone to a specifc resource in a specific enviroment in a specific project

**Comment by DeVoresyah** [comment]
hi @WcaleNieWolny can we do quick online meeting to discuss this? I'm happy to add tips for you later

**Comment by WcaleNieWolny** [comment]
yes, i would be happy to do so. Let me know when you are available

**Comment by DeVoresyah** [comment]
> yes, i would be happy to do so. Let me know when you are available

pls add me on discord `devoresyah`

**Comment by WcaleNieWolny** [comment]
I have spent some time researching and I think that i now have a good grasp of the issue. The problem with implementing what you would like is that there is no quick and easy way to do it. Currently you are either a member of a team or you are not. The proper way to implement this would be to add a new a few new columns into the `team_users`. You would add `project_id`, `enviroment_id` and `resource_id`. Essentially it would determine what you can access. For example, if `project_id` is set you would be ONLY able to access that project and if both `project_id` and `environment_id` are set you would be only access that specific enviroment in that specifc project. 

![image](https://github.com/user-attachments/assets/3e25a172-be90-485d-850d-3e5e2a6ab850)

**Comment by peaklabs-dev** [comment]
I will be working on this during the next week, as this should be included in the v4 stable (after a few priority bugs are fixed). 

The plan is as follows:
- Owner: Can do everything
- Admin: Can manage all resources, but not users, servers
- Collaborator: Can only manage the resources he created
- Custom: Add specific permissions to this role. Checkboxes with permissions and a dropdown where resources can be added to a user (services or applications).

Is this the way you would want me to implement this? Any additional things you would like to see?

**Comment by DeVoresyah** [comment]
> I will be working on this during the next week, as this should be included in the v4 stable (after a few priority bugs are fixed).
> 
> The plan is as follows:
> 
> * Owner: Can do everything
> * Admin: Can manage all resources, but not users, servers
> * Collaborator: Can only manage the resources he created
> * Custom: Add specific permissions to this role. Checkboxes with permissions and a dropdown where resources can be added to a user (services or applications).
> 
> Is this the way you would want me to implement this? Any additional things you would like to see?

perfect

**Comment by drblue** [comment]
I’m currently teaching Frontend Development at an advanced vocational school, and I’m interested in using this to host a server for my students.

Is it possible to create separate teams so that multiple students within the same team can access, modify, and trigger deployment for the same app, without being able to access apps from other groups?

Additionally, it would be great if I could set up apps for them that connect to private GitHub repositories within my organization (we use GitHub Classroom) to ensure they can’t access each other’s source code.

**Comment by gabrielcossette** [comment]
Looking forward to this!

It would solve this feature request as well (I believe?): https://github.com/coollabsio/coolify/discussions/1820

**Comment by drblue** [comment]
Any news about this feature? My API-development course where I would use this starts right after New Year’s and in February (if at all possible) I’d have to have a server running for my students 😊

**Comment by amirhmoradi** [comment]
Hi, for everyone looking for a granular permission management system on coolify v4.x :
I did a PR on coolify to solve this, which was rejected as this is supposed to be for v5. So I decided to create my own repo with the solution where it would add the feature in coolify v4 without breaking anything and having painless revert back feature. Here is the screenshot of the result and the link to the solution:

https://github.com/amirhmoradi/coolify-granular-permissions

<img width="2378" height="1964" alt="Image" src="https://github.com/user-attachments/assets/ec202003-12d6-47c3-bf88-f7f1adf4eed9" />

**Comment by tfbecker** [comment]
i am a sponsor - happy to sponsor more - someone paid for a bounty - i believe this is a well sought after feature. 
any way we can pay for prioritization? 
i love coolify and this is one of the few missing features for us

**Comment by harishn1408** [comment]
### Proposed Solution

The root cause is the lack of granular permissions and user roles in the current system, which only allows team members to access the entire general system. To address this, we need to introduce a new permission system that allows admins to invite team members to specific resources. 

Implementation plan:
1. Create a new model for resource-specific team members in `models/ResourceTeamMember.js`.
2. Add a new endpoint in `controllers/ResourceController.js` to invite team members to a specific resource.
3. Update the `views/resource/team-members.hbs` template to display the new team members menu.

Example code change in `models/ResourceTeamMember.js`:
```javascript
import { Model, DataTypes } from 'sequelize';
class ResourceTeamMember extends Model {}
ResourceTeamMember.init({
  resourceId: {
    type: DataTypes.INTEGER,
    references: {
      model: 'resources',
      key: 'id'
    }
  },
  userId: {
    type: DataTypes.INTEGER,
    references: {
      model: 'users',
      key: 'id'
    }
  }
}, {
  sequelize,
  modelName: 'ResourceTeamMember'
});
```
Testing suggestions: Verify that admins can invite team members to specific resources and that team members only have access to the resources they are invited to.

---
<sub>Happy to submit a PR implementing this approach if it looks good to the maintainers.</sub>

**Comment by oscadev** [comment]
almost?

**Comment by mogja9** [comment]
Update after checking the v5 branches: I found the existing maintainer-led permission scaffold in #9362, which was not linked in this issue's development panel. I will not duplicate that work.

If there is a distinct, currently funded portion of this issue that remains open after #9362, a maintainer can identify its acceptance criteria here; otherwise please disregard this enquiry.

**Comment by Cinzya** [comment]
FYI, we stopped the bounty program with Algora. We asked them to remove issues which are still open like this one. Even if you still see a open issue on their site, payouts won't work / happen and we don't accept any bounty claims. PRs can still be made regardless but don't expect a payout from Algora.

**Comment by tfbecker** [comment]
Out of Curiosity - why did you stop working with them ?

**Comment by Cinzya** [comment]
@tfbecker There was a large userbase that let their AI agents automate writing PRs for bounties, but they then never tested those changes, so the majority of those PRs are not working and / or broken. This is quite a burden on the team, because we still review and test those ourselves. This is even more frustrating when we try ask for clarification and either don't get a response or get very verbose answers that don't appear to come from a human.

Our goal was originally to give the Coolify community something back in form of a bounty program like Algora, but if the majority of those PRs aren't even written and audited by a human, very often new users we haven't seen contributed or interacted with Coolify before, and are barely working in the first place, we don't think this is serving our original goal.

**Comment by tfbecker** [comment]
Thank you so much for taking the time answering this question for me. I was just curious about it because I was working in venture capital before and got to know the founder and thought it was a great idea to improve the get her back bounty system. Who would have known that AI is compromising this form of community contribution?


It’s also a bit of a shame because I just had to move from Coolify to Dokploy because some of the features I’m looking for are not on the near term roadmap for you guys or they just take too long.

I think finding a middle way and scaling the review and testing system could really be a benefit (from an outside pov)


Excuse spelling mistakes, I just dictated this on my iPhone

**Comment by abdushsk** [comment]
If noone has plans to work on it, pls lmk I will take it up.

**Comment by mfyll** [comment]
After the recent changes to Coolify, this is even more relevant. We need more granulated and custom roles ASAP to easier fit every teams workflow.

**Comment by ananas-wonders** [comment]
is that or pr #11502 
is a thing we expect to have in the near future ?
